src/TypeInference/GenerateConstraints/Kinds.hs

```diff
@@ -51,6 +55,259 @@ newKVar = do
+-- should neve appear
+removeKVars uv@TST.TyUniVar{}             = return uv
+removeKVars rv@TST.TyRecVar{}             = return rv                            -- TODO -- always gets kvar (
+removeKVars tyd@TST.TyData{}              = return tyd                           -- TODO -- xtors are annotated (actual kind always inhabited)
+removeKVars tycd@TST.TyCodata{}           = return tycd                          -- TODO -- xtors are annotated (actual kind always inhabited) 
+removeKVars tydr@TST.TyDataRefined{}      = return tydr                          -- TODO -- xtors are annotated 
+removeKVars tycdr@TST.TyCodataRefined{}   = return tycdr                         -- TODO -- xtors are annotated 
+removeKVars tyn@TST.TyNominal{}           = return tyn                           -- TODO -- vartys are annotated
+-- only the contained type has to be checked
+removeKVars (TST.TySyn loc pol tyn ty)    = do 
+  ty' <- removeKVars ty
+  return $ TST.TySyn loc pol tyn ty'
+removeKVars bot@TST.TyBot{}               = return bot                           -- TODO -- always gets kvar 
+removeKVars top@TST.TyTop{}               = return top                           -- TODO -- always gets kvar
+removeKVars (TST.TyUnion loc knd ty1 ty2) = return (TST.TyUnion loc knd ty1 ty2) -- TODO -- ty1 and ty2 may contain kvars
+removeKVars (TST.TyInter loc knd ty1 ty2) = return (TST.TyInter loc knd ty1 ty2) -- TODO -- ty1 and ty2 may contain kvars
+removeKVars tr@TST.TyRec{}                = return tr                            -- TODO -- recursive type and recvar may contain kvars
+-- primitive types don't contain a type
+removeKVars i64@TST.TyI64{}               = return i64
+removeKVars f64@TST.TyF64{}               = return f64
+removeKVars ch@TST.TyChar{}               = return ch
+removeKVars str@TST.TyString{}            = return str
+-- only the contained type needs to be checked
+removeKVars (TST.TyFlipPol pol ty)        = do 
+  ty' <- removeKVars ty
+  return $ TST.TyFlipPol pol ty'
+
+
+annotTy :: RST.Typ pol -> DataDeclM (TST.Typ pol)
+annotTy (RST.TySkolemVar loc pol tv) = do 
+  polyknd <- gets declKind
+  return $ TST.TySkolemVar loc pol (getKindSkolem polyknd tv) tv 
+-- uni vars should not appear in data declarations
+annotTy (RST.TyUniVar loc _ _) = throwOtherError loc ["UniVar should not appear in data declaration"]
+annotTy (RST.TyRecVar loc pol tv) = do
+  rvs <- gets boundRecVars 
+  case M.lookup tv rvs of 
+    Nothing -> throwOtherError loc ["Unbound RecVar " <> ppPrint tv <> " in Xtor"]
+    Just knd -> return $ TST.TyRecVar loc pol knd tv
+annotTy (RST.TyData loc pol xtors) = do 
+  let xtnms = map RST.sig_name xtors
+  xtorKinds <- mapM lookupXtorKind xtnms
+  let allEq = compXtorKinds xtorKinds  
+  case allEq of 
+    Nothing -> throwOtherError loc ["Not all xtors have the same return kind"]
+    Just mk -> do 
+      xtors' <- mapM (\x -> lookupXtorSig loc x pol) xtnms
+      return $ TST.TyData loc pol mk xtors' 
```

Comment:
Ist das hier wirklich richtig?
Wenn zum Beispiel im Typ `<S(<Z>)>` steht und du dann die `XtorSig` nachschlägst, steht dann danach nicht `<S(<a : CBV>)>` im Ergebnis?
Müsstest du also nicht stattdessen die Kinds durch den Lookup erhalten und diese dann in die gegebene XtorSig einbauen, anstatt einfach die komplette Signatur aus dem Lookup zu übernehmen?

Reply:
eventuell reicht es ja auch schon, das gleiche wie im Refinement Fall zu tun: Einfach `annotateXtor` auf den `XtorSig`s aufrufen?

Reply:
Tatsache kann ich sogar die Xtors nachschauen, da diese schon im environment stehen müssen, also muss sogar noch weniger berechnet werden